src/inmanta/server/services/environment_metrics_service.py

```diff
@@ -400,11 +416,17 @@ async def get_metric_value(
         self, start_interval: datetime, end_interval: datetime, connection: asyncpg.connection.Connection
     ) -> Sequence[MetricValueTimer]:
         query: str = f"""
-            SELECT count(*)  as count,environment,sum(started-requested) as compile_waiting_time
+        WITH compile_waiting_time as (
+            SELECT count(*) as count,environment,sum(started-requested)
             FROM {Compile.table_name()}
             WHERE started >= $1
             AND started < $2
             GROUP BY environment
+        ) SELECT e.id AS environment, COALESCE(compile_waiting_time.count, 0) AS count, COALESCE(compile_waiting_time.sum, INTERVAL '0' SECOND) as compile_waiting_time
```

Comment:
In this context 0 feels weird. I also see in the design "The data returned to the client should contain the value NULL for the sub-intervals for which no data is available.", though the context is slightly different here. I wonder if we should use `null` instead of 0 for all metrics (if we do, we should make sure to keep 0 for absent statuses, and only use `null` for environments without data)? For e.g. the count metrics 0 does make a lot of sense. I need to think this over.

@arnaudsjs cc

Reply:
At first glance I like your proposal.

Reply:
I think the main question here is: How do we want to represent this scenario in the graph on the web-console?

* If we want to represent this using the value 0, we should write a value 0 into the database.
* If we want to represent this using an empty spot in the graph, we should not write a 0.

Personally, I don't have a strong opinion this. But if we would go for the second approach, I am wondering whether we should write anything at all to the database. As far a I can see, we want the same behavior in bullet point 2 as when the server is offline and no data is written to the database. Do I see that correctly?

Reply:
Let me start by saying I agree with your conclusion that if we pick option 2, we don't need to write anything at all.

Now, for some metrics (e.g. any of the count metrics) I would say option 1 makes more sense: a newly created environment really does have 0 instances, resources and relevant agents. For compile rate we could argue the same. For all of those, absence naturally translates to zero. But then for compile time and compile waiting time this is not the case: adding zeros may skew front-end aggregations, potentially giving the false impression of a compile time regression.

Therefore, I would propose we extend the docstring of the method on the ABC with something along the lines of "If, at the time of collection, no metric data exists for an environment and/or group, the implementation should return a meaningful default  (e.g. 0 for count metrics). If no meaningful default exists (e.g. for some time metrics), the data may be left out."

This would allow us to make a guided choice for each metric individually. In our current case, we could leave the data out for the compile time and waiting time metrics.

What do you think?

@Hugo-Inmanta @wouterdb CC, since this is turning into a design discussion rather than just a PR review. If you need more context, let me know.

Reply:
I agree.

Reply:
this seems like to most flexible and right solution. I'll change it.

Reply:
I now realize the compile time and waiting time metrics contain a count as well, so writing zero would not have been so bad after all, since it's up to the collection endpoint to do correct aggregation anyway. Either way, I don't think leaving it out does any harm.